compiler/rustc_mir_transform/src/value_numbering.rs

```diff
@@ -0,0 +1,907 @@
+//! Performs local value numbering.
+//!
+//! Local value numbering is a powerful analysis that subsumes constant and copy propagation.  Since
+//! it can reason about immutability of memory, it can do optimizations that LLVM won't. Because
+//! it's local to a basic block, however, it doesn't handle all the cases that our constant and copy
+//! propagation passes do. Eventually, when we have SSA, we should be able to upgrade this to global
+//! value numbering.
+
+use crate::simplify;
+use crate::ty::{Region, Ty};
+use crate::MirPass;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::visit::{MutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::{
+    AggregateKind, BasicBlock, BasicBlockData, BinOp, Body, BorrowKind, CastKind, Constant,
+    InlineAsmOperand, Local, LocalDecls, Location, Mutability, NullOp, Operand, Place, PlaceElem,
+    ProjectionElem, Rvalue, SourceInfo, Statement, StatementKind, TerminatorKind, UnOp,
+};
+use rustc_middle::ty::{self, Const, ParamEnv, TyCtxt};
+use rustc_span::def_id::DefId;
+use std::collections::hash_map::Entry;
+use std::mem;
+
+/// Performs copy and constant propagation via value numbering.
+pub struct ValueNumbering;
+
+impl<'tcx> MirPass<'tcx> for ValueNumbering {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        // We run on MIR opt level 2 or greater, if `-Z number-values` is passed.
+        //
+        // FIXME(pcwalton): Enable by default.
+        sess.mir_opt_level() >= 2 && sess.opts.debugging_opts.number_values
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // First, promote mutable locals to immutable ones. This helps our rudimentary alias
+        // analysis a lot.
```

Comment:
The precedent for indirect mutation is to traverse the MIR to looking for any locals that are mutably borrowed at any point. I'm not sure if you need to look at the mutability of a LocalDecl separately, but maybe doing it that way helps you avoid dealing with that failing test case?

Reply:
I'm not sure what you mean, could you elaborate?

Reply:
Sure, see [this comment for contxt](https://github.com/rust-lang/rust/pull/92051#discussion_r771748033). My understanding is that you're using the mutability field of each `LocalDecl` as a conservative approximation for whether it can be mutated indirectly (through a pointer). Is that correct? What you're doing is fine of course, but it would be slightly more precise to traverse the MIR looking for any local that is the base of an `Rvalue::Ref(Mut)` or `Rvalue::AddrOf`. I don't know how much the extra precision will matter.

Reply:
Well, actually is what you're doing fine? Is your code correct in the presence of locals with interior mutability?

Reply:
Look at the code to handle `UnsafeCell`.

Reply:
Most of this concern is resolved by the fact that we don't have to care about function calls. One more nit: could you explain what would happen in this (incredibly niche) case?

```rust
//! I am in libcore/src/cell.rs, so I have access to `UnsafeCell`'s field.

let x  = UnsafeCell { value: 1 }; // `x` is not marked as a mutable local (I think)?
let px = &x;
unsafe { *(px as *mut _).value  } = 2 };
read(x.value);
```

I think you'll handle this correctly due to several conservative assumptions you've made. `UnsafeCell` is not `Copy`, for one, and also you bail out on field projections. Is this accurate or is something more robust? 

Frankly, I suspect there are other MIR transformations that are unsound if  direct accesses to the field of an `UnsafeCell` are allowed, but it doesn't hurt to ask I guess.